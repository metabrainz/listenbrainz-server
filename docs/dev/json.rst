.. _json-doc:

JSON Documentation
==================

.. note:: Do not submit copyrighted information in these fields!

Submission JSON
---------------

To submit a listen via our API (see: :doc:`api`), ``POST`` a JSON document to
the ``submit-listens`` endpoint. Submit one of three types JSON documents:

- ``single``: Submit single listen

   - Indicates user just finished listening to track

   - ``payload`` should contain information about *exactly one* track

- ``playing_now``: Submit ``playing_now`` notification

   - Indicates that user just began listening to track

   - ``payload`` should contain information about *exactly one* track

   - Submitting ``playing_now`` documents is optional

   - Timestamp must be omitted from a ``playing_now`` submission.

- ``import``: Submit previously saved listens

   - ``payload`` should contain information about *at least one* track

   - submitting multiple listens in one go is allowed, but the complete JSON
     document may not exceed :data:`~webserver.views.api.MAX_LISTEN_SIZE` bytes
     in size

The ``listen_type`` element defines different types of submissions. The element
is placed at the top-most level of the JSON document. The only other required
element is the ``payload`` element. This provides an array of listens â€“ the
payload may be one or more listens (as designated by ``listen_type``)::

    {
      "listen_type": "single",
      "payload": [
          --- listen data here ---
      ]
    }

A sample listen payload may look like::

    {
      "listened_at": 1443521965,
      "track_metadata": {
        "additional_info": {
          "release_mbid": "bf9e91ea-8029-4a04-a26a-224e00a83266",
          "artist_mbids": [
            "db92a151-1ac2-438b-bc43-b82e149ddd50"
          ],
          "recording_mbid": "98255a8c-017a-4bc7-8dd6-1fa36124572b",
          "tags": [ "you", "just", "got", "rick rolled!"]
        },
        "artist_name": "Rick Astley",
        "track_name": "Never Gonna Give You Up",
        "release_name": "Whenever you need somebody"
      }
    }

A complete submit listen JSON document may look like::

    {
      "listen_type": "single",
      "payload": [
        {
          "listened_at": 1443521965,
          "track_metadata": {
            "additional_info": {
              "listening_from": "spotify",
              "release_mbid": "bf9e91ea-8029-4a04-a26a-224e00a83266",
              "artist_mbids": [
                "db92a151-1ac2-438b-bc43-b82e149ddd50"
              ],
              "recording_mbid": "98255a8c-017a-4bc7-8dd6-1fa36124572b",
              "tags": [ "you", "just", "got", "rick rolled!"]
            },
            "artist_name": "Rick Astley",
            "track_name": "Never Gonna Give You Up",
            "release_name": "Whenever you need somebody"
          }
        }
      ]
    }


Fetching listen JSON
--------------------

The JSON documents returned from our API look like the following::

    {
      "payload": {
        "count": 25,
        "user_id": "-- the MusicBrainz ID of the user --",
        "listens": [
          "-- listen data here ---"
        ]
      }
    }

The number of listens in the document are returned by the top-level ``count``
element. The ``user_id`` element contains the MusicBrainz ID of the user whose listens are
being returned. The other element is the ``listens`` element. This is a list which contains
the listen JSON elements (described above).

The JSON document returned by the API endpoint for getting tracks being played right now
is the same as above, except that it also contains the ``payload/playing_now`` element as a
boolean set to True.


Payload JSON details
--------------------

A minimal payload must include
``track_metadata/artist_name`` and ``track_metadata/track_name`` elements::

    {
      "track_metadata": {
        "artist_name": "Rick Astley",
        "track_name": "Never Gonna Give You Up",
      }
    }

``artist_name`` and ``track_name`` elements must be simple strings.

The payload will also include the ``listened_at`` element which must be an integer
representing the Unix time when the track was listened to. The only exception to this
rule is when the listen is being played right now and has been retrieved from the
endpoint to get listens being played right now. The ``listened_at`` element will be
absent for such listens.

Add additional metadata you may have for a track to the ``additional_info``
element. Any additional information allows us to better correlate your listen
data to existing MusicBrainz-based data. If you have MusicBrainz IDs available,
submit them!

The following optional elements may also be included in the ``track_metadata`` element:

======================= ===========================================================================================================================================================================================================================================================================================================================================================================================================
element                 description
======================= ===========================================================================================================================================================================================================================================================================================================================================================================================================
``release_name``        the name of the release this recording was played from.
======================= ===========================================================================================================================================================================================================================================================================================================================================================================================================

The following optional elements may also be included in the ``additional_info`` element. If you do not have
the data for any of the following fields, omit the key entirely:

======================= ===========================================================================================================================================================================================================================================================================================================================================================================================================
element                 description
======================= ===========================================================================================================================================================================================================================================================================================================================================================================================================
``artist_mbids``        A list of MusicBrainz Artist IDs, one or more Artist IDs may be included here. If you have a complete MusicBrainz artist credit that contains multiple Artist IDs, include them all in this list.
``release_group_mbid``  A MusicBrainz Release Group ID of the release group this recording was played from.
``release_mbid``        A MusicBrainz Release ID of the release this recording was played from.
``recording_mbid``      A MusicBrainz Recording ID of the recording that was played.
``track_mbid``          A MusicBrainz Track ID associated with the recording that was played.
``work_mbids``          A list of MusicBrainz Work IDs that may be associated with this recording.
``tracknumber``         The tracknumber of the recording. This first recording on a release is tracknumber 1.
``isrc``                The ISRC code associated with the recording.
``spotify_id``          The Spotify track URL associated with this recording.  e.g.: http://open.spotify.com/track/1rrgWMXGCGHru5bIRxGFV0
``tags``                A list of user defined tags to be associated with this recording. These tags are similar to last.fm tags. For example, you have apply tags such as ``punk``, ``see-live``, ``smelly``. You may submit up to :data:`~webserver.views.api.MAX_TAGS_PER_LISTEN` tags and each tag may be up to :data:`~webserver.views.api.MAX_TAG_SIZE` characters large.
``listening_from``       The source of the listen, i.e the name of the client or service which submits the listen.
======================= ===========================================================================================================================================================================================================================================================================================================================================================================================================

At this point, we are not scrubbing any superflous elements that may be
submitted via the ``additional_info`` element. We're open to see how people
will make use of these unspecified fields and may decide to formally specify or
scrub elements in the future.
