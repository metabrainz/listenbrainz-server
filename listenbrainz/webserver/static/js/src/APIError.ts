export default class APIError extends Error {
  status?: string;

  response?: Response;
}
