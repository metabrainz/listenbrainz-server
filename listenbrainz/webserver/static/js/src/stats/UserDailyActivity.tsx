import * as React from "react";
import { faExclamationCircle, faLink } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { IconProp } from "@fortawesome/fontawesome-svg-core";

import APIService from "../APIService";
import Card from "../components/Card";
import HeatMap from "./HeatMap";
import Loader from "../components/Loader";

export type UserDailyActivityProps = {
  range: UserStatsAPIRange;
  user: ListenBrainzUser;
  apiUrl: string;
};

export type UserDailyActivityState = {
  data: UserDailyActivityData;
  loading: boolean;
  graphContainerWidth?: number;
  errorMessage: string;
  hasError: boolean;
};

export default class UserDailyActivity extends React.Component<
  UserDailyActivityProps,
  UserDailyActivityState
> {
  APIService: APIService;

  graphContainer: React.RefObject<HTMLDivElement>;

  constructor(props: UserDailyActivityProps) {
    super(props);

    this.APIService = new APIService(
      props.apiUrl || `${window.location.origin}/1`
    );

    this.state = {
      data: [],
      loading: false,
      errorMessage: "",
      hasError: false,
    };

    this.graphContainer = React.createRef();
  }

  componentDidMount() {
    window.addEventListener("resize", this.handleResize);

    this.handleResize();
  }

  componentDidUpdate(prevProps: UserDailyActivityProps) {
    const { range: prevRange } = prevProps;
    const { range: currRange } = this.props;
    if (prevRange !== currRange) {
      if (["week", "month", "year", "all_time"].indexOf(currRange) < 0) {
        this.setState({
          loading: false,
          hasError: true,
          errorMessage: `Invalid range: ${currRange}`,
        });
      } else {
        this.loadData();
      }
    }
  }

  componentWillUnmount() {
    window.removeEventListener("resize", this.handleResize);
  }

  getData = async (): Promise<UserDailyActivityResponse> => {
    const { range, user } = this.props;
    try {
      const data = await this.APIService.getUserDailyActivity(user.name, range);
      return data;
    } catch (error) {
      if (error.response && error.response.status === 204) {
        this.setState({
          loading: false,
          hasError: true,
          errorMessage: "Statistics for the user have not been calculated",
        });
      } else {
        this.setState(() => {
          throw error;
        });
      }
    }
    return {} as UserDailyActivityResponse;
  };

  loadData = async (): Promise<void> => {
    this.setState({
      hasError: false,
      loading: true,
    });
    const data = await this.getData();
    this.setState({
      data: this.processData(data),
      loading: false,
    });
  };

  processData = (data: UserDailyActivityResponse): UserDailyActivityData => {
    const weekdays = [
      "Monday",
      "Tuesday",
      "Wednesday",
      "Thursday",
      "Friday",
      "Saturday",
      "Sunday",
    ];

    const result: UserDailyActivityData = [];

    const tzOffset = -Math.floor(new Date().getTimezoneOffset() / 60);

    weekdays.forEach((day) => {
      const dayData = data.payload.daily_activity[day];
      const hourData: any = {};

      dayData.forEach((elem) => {
        const hour = (elem.hour + tzOffset + 24) % 24;
        hourData[hour] = elem.listen_count;
      });

      result.push({
        day,
        ...hourData,
      });
    });

    const average = Array(24).fill(0);
    Object.values(data.payload.daily_activity).forEach((dayData) => {
      dayData.forEach((hourData) => {
        average[hourData.hour] += hourData.listen_count;
      });
    });

    const averageData: any = {};
    average.forEach((elem, index) => {
      const hour = (index + tzOffset + 24) % 24;
      averageData[hour] = Math.ceil(elem / 7);
    });

    result.unshift({
      day: "Average",
      ...averageData,
    });

    return result;
  };

  handleResize = () => {
    this.setState({
      graphContainerWidth: this.graphContainer.current?.offsetWidth,
    });
  };

  render() {
    const {
      data,
      loading,
      graphContainerWidth,
      hasError,
      errorMessage,
    } = this.state;

    return (
      <Card
        style={{
          marginTop: 20,
          minHeight: (graphContainerWidth || 1200) * 0.4,
        }}
        ref={this.graphContainer}
      >
        <div className="row">
          <div className="col-xs-10">
            <h3 className="capitalize-bold" style={{ marginLeft: 20 }}>
              Daily Activity
            </h3>
          </div>
          <div className="col-xs-2 text-right">
            <h4 style={{ marginTop: 20 }}>
              <a href="#daily-activity">
                <FontAwesomeIcon
                  icon={faLink as IconProp}
                  size="sm"
                  color="#000000"
                  style={{ marginRight: 20 }}
                />
              </a>
            </h4>
          </div>
        </div>
        <Loader isLoading={loading}>
          {hasError && (
            <div
              style={{
                display: "flex",
                alignItems: "center",
                justifyContent: "center",
                minHeight: "inherit",
              }}
            >
              <span style={{ fontSize: 24 }}>
                <FontAwesomeIcon icon={faExclamationCircle as IconProp} />{" "}
                {errorMessage}
              </span>
            </div>
          )}
          {!hasError && (
            <div className="row">
              <div className="col-xs-12">
                {graphContainerWidth && (
                  <HeatMap data={data} width={graphContainerWidth} />
                )}
              </div>
            </div>
          )}
        </Loader>
      </Card>
    );
  }
}
