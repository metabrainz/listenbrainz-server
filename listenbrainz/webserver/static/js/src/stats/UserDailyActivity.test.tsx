import * as React from "react";
import { mount, shallow } from "enzyme";

import UserDailyActivity, { UserDailyActivityProps } from "./UserDailyActivity";
import APIError from "../APIError";
import * as userDailyActivityResponse from "../__mocks__/userDailyActivity.json";
import * as userDailyActivityProcessedData from "../__mocks__/userDailyActivityProcessData.json";

const props: UserDailyActivityProps = {
  user: {
    name: "foobar",
  },
  range: "week",
  apiUrl: "barfoo",
};

// Set timeZone to UTC+5:30 because the testdata is in that format
// eslint-disable-next-line no-extend-native
Date.prototype.getTimezoneOffset = () => -330;

describe("UserDailyActivity", () => {
  it("renders correctly", () => {
    const wrapper = shallow<UserDailyActivity>(
      <UserDailyActivity {...{ ...props, range: "all_time" }} />
    );

    wrapper.setState({
      data: userDailyActivityProcessedData,
      graphContainerWidth: 1200,
      loading: false,
    });
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
  });

  it("renders corectly when range is invalid", () => {
    const wrapper = mount<UserDailyActivity>(<UserDailyActivity {...props} />);

    wrapper.setProps({ range: "invalid_range" as UserStatsAPIRange });
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
  });
});

describe("componentDidMount", () => {
  it('adds event listener for "resize" event', () => {
    const wrapper = shallow<UserDailyActivity>(
      <UserDailyActivity {...props} />
    );
    const instance = wrapper.instance();

    const spy = jest.spyOn(window, "addEventListener");
    spy.mockImplementationOnce(() => {});
    instance.handleResize = jest.fn();
    instance.componentDidMount();

    expect(spy).toHaveBeenCalledWith("resize", instance.handleResize);
  });

  it('calls "handleResize" once', () => {
    const wrapper = shallow<UserDailyActivity>(
      <UserDailyActivity {...props} />
    );
    const instance = wrapper.instance();

    instance.handleResize = jest.fn();
    instance.componentDidMount();

    expect(instance.handleResize).toHaveBeenCalledTimes(1);
  });
});

describe("componentDidUpdate", () => {
  it("it sets correct state if range is incorrect", () => {
    const wrapper = shallow<UserDailyActivity>(
      <UserDailyActivity {...props} />
    );

    wrapper.setProps({ range: "invalid_range" as UserStatsAPIRange });
    wrapper.update();

    expect(wrapper.state()).toMatchObject({
      loading: false,
      hasError: true,
      errorMessage: "Invalid range: invalid_range",
    });
  });

  it("calls loadData once if range is valid", () => {
    const wrapper = shallow<UserDailyActivity>(
      <UserDailyActivity {...props} />
    );
    const instance = wrapper.instance();

    instance.loadData = jest.fn();
    wrapper.setProps({ range: "month" });
    wrapper.update();

    expect(instance.loadData).toHaveBeenCalledTimes(1);
  });
});

describe("componentWillUnmount", () => {
  it('removes event listener for "resize" event', () => {
    const wrapper = shallow<UserDailyActivity>(
      <UserDailyActivity {...props} />
    );
    const instance = wrapper.instance();

    const spy = jest.spyOn(window, "removeEventListener");
    spy.mockImplementationOnce(() => {});
    instance.handleResize = jest.fn();
    instance.componentWillUnmount();

    expect(spy).toHaveBeenCalledWith("resize", instance.handleResize);
  });
});

describe("getData", () => {
  it("calls getUserDailyActivity with correct params", async () => {
    const wrapper = shallow<UserDailyActivity>(
      <UserDailyActivity {...props} />
    );
    const instance = wrapper.instance();

    const spy = jest.spyOn(instance.APIService, "getUserDailyActivity");
    spy.mockImplementation(() =>
      Promise.resolve(userDailyActivityResponse as UserDailyActivityResponse)
    );
    const result = await instance.getData();

    expect(spy).toHaveBeenCalledWith("foobar", "week");
    expect(result).toEqual(userDailyActivityResponse);
  });

  it("sets state correctly if data is not calculated", async () => {
    const wrapper = shallow<UserDailyActivity>(
      <UserDailyActivity {...props} />
    );
    const instance = wrapper.instance();

    const spy = jest.spyOn(instance.APIService, "getUserDailyActivity");
    const noContentError = new APIError("NO CONTENT");
    noContentError.response = {
      status: 204,
    } as Response;
    spy.mockImplementation(() => Promise.reject(noContentError));
    await instance.getData();

    expect(wrapper.state()).toMatchObject({
      loading: false,
      hasError: true,
      errorMessage: "Statistics for the user have not been calculated",
    });
  });

  it("throws error", async () => {
    const wrapper = shallow<UserDailyActivity>(
      <UserDailyActivity {...props} />
    );
    const instance = wrapper.instance();

    const spy = jest.spyOn(instance.APIService, "getUserDailyActivity");
    const notFoundError = new APIError("NOT FOUND");
    notFoundError.response = {
      status: 404,
    } as Response;
    spy.mockImplementation(() => Promise.reject(notFoundError));

    await expect(instance.getData()).rejects.toThrow("NOT FOUND");
  });
});

describe("processData", () => {
  it("processes data correctly for all_time", () => {
    const wrapper = shallow<UserDailyActivity>(
      <UserDailyActivity {...{ ...props, range: "all_time" }} />
    );
    const instance = wrapper.instance();

    const result = instance.processData(
      userDailyActivityResponse as UserDailyActivityResponse
    );

    expect(result).toEqual(userDailyActivityProcessedData);
  });
});

describe("loadData", () => {
  it("calls getData once", async () => {
    const wrapper = shallow<UserDailyActivity>(
      <UserDailyActivity {...props} />
    );
    const instance = wrapper.instance();

    instance.getData = jest.fn();
    instance.processData = jest.fn();
    await instance.loadData();

    expect(instance.getData).toHaveBeenCalledTimes(1);
  });

  it("set state correctly", async () => {
    const wrapper = shallow<UserDailyActivity>(
      <UserDailyActivity {...props} />
    );
    const instance = wrapper.instance();

    instance.getData = jest
      .fn()
      .mockImplementationOnce(() => Promise.resolve(userDailyActivityResponse));
    await instance.loadData();

    expect(wrapper.state()).toMatchObject({
      data: userDailyActivityProcessedData,
      loading: false,
    });
  });
});
