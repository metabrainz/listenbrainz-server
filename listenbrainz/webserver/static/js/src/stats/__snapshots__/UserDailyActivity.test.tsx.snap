// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDailyActivity renders corectly when range is invalid 1`] = `
<UserDailyActivity
  apiUrl="barfoo"
  range="invalid_range"
  user={
    Object {
      "name": "foobar",
    }
  }
>
  <ForwardRef
    style={
      Object {
        "marginTop": 20,
        "minHeight": 480,
      }
    }
  >
    <div
      style={
        Object {
          "background": "#FFFFFF",
          "border": "1px solid #EEEEEE",
          "borderRadius": "8px",
          "boxShadow": "0 1px 1px rgba(192,192,192,0.1), 0 2px 2px rgba(192,192,192,0.15), 0 4px 4px rgba(192,192,192,0.20)",
          "boxSizing": "border-box",
          "height": "100%",
          "marginTop": 20,
          "minHeight": 480,
        }
      }
    >
      <div
        className="row"
      >
        <div
          className="col-xs-10"
        >
          <h3
            className="capitalize-bold"
            style={
              Object {
                "marginLeft": 20,
              }
            }
          >
            Daily Activity
          </h3>
        </div>
        <div
          className="col-xs-2 text-right"
        >
          <h4
            style={
              Object {
                "marginTop": 20,
              }
            }
          >
            <a
              href="#daily-activity"
            >
              <FontAwesomeIcon
                border={false}
                className=""
                color="#000000"
                fixedWidth={false}
                flip={null}
                icon={
                  Object {
                    "icon": Array [
                      512,
                      512,
                      Array [],
                      "f0c1",
                      "M326.612 185.391c59.747 59.809 58.927 155.698.36 214.59-.11.12-.24.25-.36.37l-67.2 67.2c-59.27 59.27-155.699 59.262-214.96 0-59.27-59.26-59.27-155.7 0-214.96l37.106-37.106c9.84-9.84 26.786-3.3 27.294 10.606.648 17.722 3.826 35.527 9.69 52.721 1.986 5.822.567 12.262-3.783 16.612l-13.087 13.087c-28.026 28.026-28.905 73.66-1.155 101.96 28.024 28.579 74.086 28.749 102.325.51l67.2-67.19c28.191-28.191 28.073-73.757 0-101.83-3.701-3.694-7.429-6.564-10.341-8.569a16.037 16.037 0 0 1-6.947-12.606c-.396-10.567 3.348-21.456 11.698-29.806l21.054-21.055c5.521-5.521 14.182-6.199 20.584-1.731a152.482 152.482 0 0 1 20.522 17.197zM467.547 44.449c-59.261-59.262-155.69-59.27-214.96 0l-67.2 67.2c-.12.12-.25.25-.36.37-58.566 58.892-59.387 154.781.36 214.59a152.454 152.454 0 0 0 20.521 17.196c6.402 4.468 15.064 3.789 20.584-1.731l21.054-21.055c8.35-8.35 12.094-19.239 11.698-29.806a16.037 16.037 0 0 0-6.947-12.606c-2.912-2.005-6.64-4.875-10.341-8.569-28.073-28.073-28.191-73.639 0-101.83l67.2-67.19c28.239-28.239 74.3-28.069 102.325.51 27.75 28.3 26.872 73.934-1.155 101.96l-13.087 13.087c-4.35 4.35-5.769 10.79-3.783 16.612 5.864 17.194 9.042 34.999 9.69 52.721.509 13.906 17.454 20.446 27.294 10.606l37.106-37.106c59.271-59.259 59.271-155.699.001-214.959z",
                    ],
                    "iconName": "link",
                    "prefix": "fas",
                  }
                }
                inverse={false}
                listItem={false}
                mask={null}
                pull={null}
                pulse={false}
                rotation={null}
                size="sm"
                spin={false}
                style={
                  Object {
                    "marginRight": 20,
                  }
                }
                swapOpacity={false}
                symbol={false}
                title=""
                transform={null}
              >
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-link fa-w-16 fa-sm "
                  color="#000000"
                  data-icon="link"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  style={
                    Object {
                      "marginRight": 20,
                    }
                  }
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M326.612 185.391c59.747 59.809 58.927 155.698.36 214.59-.11.12-.24.25-.36.37l-67.2 67.2c-59.27 59.27-155.699 59.262-214.96 0-59.27-59.26-59.27-155.7 0-214.96l37.106-37.106c9.84-9.84 26.786-3.3 27.294 10.606.648 17.722 3.826 35.527 9.69 52.721 1.986 5.822.567 12.262-3.783 16.612l-13.087 13.087c-28.026 28.026-28.905 73.66-1.155 101.96 28.024 28.579 74.086 28.749 102.325.51l67.2-67.19c28.191-28.191 28.073-73.757 0-101.83-3.701-3.694-7.429-6.564-10.341-8.569a16.037 16.037 0 0 1-6.947-12.606c-.396-10.567 3.348-21.456 11.698-29.806l21.054-21.055c5.521-5.521 14.182-6.199 20.584-1.731a152.482 152.482 0 0 1 20.522 17.197zM467.547 44.449c-59.261-59.262-155.69-59.27-214.96 0l-67.2 67.2c-.12.12-.25.25-.36.37-58.566 58.892-59.387 154.781.36 214.59a152.454 152.454 0 0 0 20.521 17.196c6.402 4.468 15.064 3.789 20.584-1.731l21.054-21.055c8.35-8.35 12.094-19.239 11.698-29.806a16.037 16.037 0 0 0-6.947-12.606c-2.912-2.005-6.64-4.875-10.341-8.569-28.073-28.073-28.191-73.639 0-101.83l67.2-67.19c28.239-28.239 74.3-28.069 102.325.51 27.75 28.3 26.872 73.934-1.155 101.96l-13.087 13.087c-4.35 4.35-5.769 10.79-3.783 16.612 5.864 17.194 9.042 34.999 9.69 52.721.509 13.906 17.454 20.446 27.294 10.606l37.106-37.106c59.271-59.259 59.271-155.699.001-214.959z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
              </FontAwesomeIcon>
            </a>
          </h4>
        </div>
      </div>
      <Loader
        isLoading={false}
      >
        <div
          style={
            Object {
              "alignItems": "center",
              "display": "flex",
              "justifyContent": "center",
              "minHeight": "inherit",
            }
          }
        >
          <span
            style={
              Object {
                "fontSize": 24,
              }
            }
          >
            <FontAwesomeIcon
              border={false}
              className=""
              fixedWidth={false}
              flip={null}
              icon={
                Object {
                  "icon": Array [
                    512,
                    512,
                    Array [],
                    "f06a",
                    "M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z",
                  ],
                  "iconName": "exclamation-circle",
                  "prefix": "fas",
                }
              }
              inverse={false}
              listItem={false}
              mask={null}
              pull={null}
              pulse={false}
              rotation={null}
              size={null}
              spin={false}
              swapOpacity={false}
              symbol={false}
              title=""
              transform={null}
            >
              <svg
                aria-hidden="true"
                className="svg-inline--fa fa-exclamation-circle fa-w-16 "
                data-icon="exclamation-circle"
                data-prefix="fas"
                focusable="false"
                role="img"
                style={Object {}}
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
                  fill="currentColor"
                  style={Object {}}
                />
              </svg>
            </FontAwesomeIcon>
             
            Invalid range: invalid_range
          </span>
        </div>
      </Loader>
    </div>
  </ForwardRef>
</UserDailyActivity>
`;

exports[`UserDailyActivity renders correctly 1`] = `
<ForwardRef
  style={
    Object {
      "marginTop": 20,
      "minHeight": 480,
    }
  }
>
  <div
    className="row"
  >
    <div
      className="col-xs-10"
    >
      <h3
        className="capitalize-bold"
        style={
          Object {
            "marginLeft": 20,
          }
        }
      >
        Daily Activity
      </h3>
    </div>
    <div
      className="col-xs-2 text-right"
    >
      <h4
        style={
          Object {
            "marginTop": 20,
          }
        }
      >
        <a
          href="#daily-activity"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            color="#000000"
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  512,
                  512,
                  Array [],
                  "f0c1",
                  "M326.612 185.391c59.747 59.809 58.927 155.698.36 214.59-.11.12-.24.25-.36.37l-67.2 67.2c-59.27 59.27-155.699 59.262-214.96 0-59.27-59.26-59.27-155.7 0-214.96l37.106-37.106c9.84-9.84 26.786-3.3 27.294 10.606.648 17.722 3.826 35.527 9.69 52.721 1.986 5.822.567 12.262-3.783 16.612l-13.087 13.087c-28.026 28.026-28.905 73.66-1.155 101.96 28.024 28.579 74.086 28.749 102.325.51l67.2-67.19c28.191-28.191 28.073-73.757 0-101.83-3.701-3.694-7.429-6.564-10.341-8.569a16.037 16.037 0 0 1-6.947-12.606c-.396-10.567 3.348-21.456 11.698-29.806l21.054-21.055c5.521-5.521 14.182-6.199 20.584-1.731a152.482 152.482 0 0 1 20.522 17.197zM467.547 44.449c-59.261-59.262-155.69-59.27-214.96 0l-67.2 67.2c-.12.12-.25.25-.36.37-58.566 58.892-59.387 154.781.36 214.59a152.454 152.454 0 0 0 20.521 17.196c6.402 4.468 15.064 3.789 20.584-1.731l21.054-21.055c8.35-8.35 12.094-19.239 11.698-29.806a16.037 16.037 0 0 0-6.947-12.606c-2.912-2.005-6.64-4.875-10.341-8.569-28.073-28.073-28.191-73.639 0-101.83l67.2-67.19c28.239-28.239 74.3-28.069 102.325.51 27.75 28.3 26.872 73.934-1.155 101.96l-13.087 13.087c-4.35 4.35-5.769 10.79-3.783 16.612 5.864 17.194 9.042 34.999 9.69 52.721.509 13.906 17.454 20.446 27.294 10.606l37.106-37.106c59.271-59.259 59.271-155.699.001-214.959z",
                ],
                "iconName": "link",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size="sm"
            spin={false}
            style={
              Object {
                "marginRight": 20,
              }
            }
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
        </a>
      </h4>
    </div>
  </div>
  <Loader
    isLoading={false}
  >
    <div
      className="row"
    >
      <div
        className="col-xs-12"
      >
        <Heatmap
          data={
            Array [
              Object {
                "0": 16,
                "1": 8,
                "10": 39,
                "11": 51,
                "12": 59,
                "13": 33,
                "14": 34,
                "15": 39,
                "16": 60,
                "17": 68,
                "18": 39,
                "19": 51,
                "2": 1,
                "20": 32,
                "21": 50,
                "22": 37,
                "23": 26,
                "3": 0,
                "4": 0,
                "5": 0,
                "6": 0,
                "7": 5,
                "8": 20,
                "9": 25,
                "day": "Average",
              },
              Object {
                "0": 19,
                "1": 12,
                "10": 56,
                "11": 60,
                "12": 71,
                "13": 42,
                "14": 30,
                "15": 46,
                "16": 72,
                "17": 52,
                "18": 41,
                "19": 58,
                "2": 2,
                "20": 37,
                "21": 63,
                "22": 48,
                "23": 32,
                "3": 0,
                "4": 0,
                "5": 0,
                "6": 0,
                "7": 1,
                "8": 35,
                "9": 36,
                "day": "Monday",
              },
              Object {
                "0": 19,
                "1": 8,
                "10": 50,
                "11": 33,
                "12": 46,
                "13": 33,
                "14": 44,
                "15": 42,
                "16": 67,
                "17": 60,
                "18": 50,
                "19": 56,
                "2": 0,
                "20": 28,
                "21": 34,
                "22": 47,
                "23": 14,
                "3": 0,
                "4": 0,
                "5": 0,
                "6": 0,
                "7": 0,
                "8": 17,
                "9": 60,
                "day": "Tuesday",
              },
              Object {
                "0": 2,
                "1": 1,
                "10": 45,
                "11": 47,
                "12": 82,
                "13": 45,
                "14": 14,
                "15": 37,
                "16": 52,
                "17": 64,
                "18": 54,
                "19": 63,
                "2": 0,
                "20": 29,
                "21": 56,
                "22": 26,
                "23": 30,
                "3": 0,
                "4": 0,
                "5": 0,
                "6": 0,
                "7": 0,
                "8": 29,
                "9": 20,
                "day": "Wednesday",
              },
              Object {
                "0": 28,
                "1": 7,
                "10": 14,
                "11": 62,
                "12": 85,
                "13": 21,
                "14": 45,
                "15": 57,
                "16": 84,
                "17": 79,
                "18": 29,
                "19": 50,
                "2": 0,
                "20": 33,
                "21": 57,
                "22": 33,
                "23": 21,
                "3": 0,
                "4": 0,
                "5": 0,
                "6": 0,
                "7": 16,
                "8": 16,
                "9": 17,
                "day": "Thursday",
              },
              Object {
                "0": 8,
                "1": 15,
                "10": 53,
                "11": 57,
                "12": 39,
                "13": 31,
                "14": 48,
                "15": 45,
                "16": 79,
                "17": 133,
                "18": 56,
                "19": 39,
                "2": 1,
                "20": 31,
                "21": 43,
                "22": 51,
                "23": 26,
                "3": 0,
                "4": 0,
                "5": 0,
                "6": 0,
                "7": 5,
                "8": 30,
                "9": 30,
                "day": "Friday",
              },
              Object {
                "0": 20,
                "1": 12,
                "10": 26,
                "11": 44,
                "12": 56,
                "13": 29,
                "14": 28,
                "15": 23,
                "16": 25,
                "17": 55,
                "18": 18,
                "19": 57,
                "2": 1,
                "20": 37,
                "21": 80,
                "22": 29,
                "23": 22,
                "3": 0,
                "4": 0,
                "5": 0,
                "6": 0,
                "7": 0,
                "8": 7,
                "9": 11,
                "day": "Saturday",
              },
              Object {
                "0": 10,
                "1": 0,
                "10": 26,
                "11": 52,
                "12": 30,
                "13": 24,
                "14": 25,
                "15": 17,
                "16": 36,
                "17": 30,
                "18": 24,
                "19": 29,
                "2": 0,
                "20": 28,
                "21": 16,
                "22": 24,
                "23": 35,
                "3": 0,
                "4": 0,
                "5": 0,
                "6": 0,
                "7": 8,
                "8": 2,
                "9": 0,
                "day": "Sunday",
              },
            ]
          }
          width={1200}
        />
      </div>
    </div>
  </Loader>
</ForwardRef>
`;
